### ISSUE TITLE
AI (Internal): [Azure-Core] ERROR: Exception in Command Processing for EventSource Azure-Core: An instance of EventSource with Guid xxx already exists

### ISSUE THREAD

**Issue by MikePennington** [Azure/azure-functions-host#7393]
We have multiple function apps that all started receiving this error in the logs starting around 1:00PM Central Time on 5/17.

```
AI (Internal): [Azure-Core] ERROR: Exception in Command Processing for EventSource Azure-Core: An instance of EventSource with Guid 44cbc7c6-6776-5f3c-36c1-75cd3ef19ea9 already exists.
```

At this point I am unsure if this error is negatively impacting the functions in any way, but I am investigating further.

We use functions quite a bit, and about half of them starting reporting this error at this time. Anecdotally it seems to be affecting our longer-running functions and not are shorter-running functions. This is affecting functions of multiple types: queue-triggered and timer-triggered functions, and http-triggered functions. We are not creating a new TelemetryClient anywhere in our code (https://github.com/microsoft/ApplicationInsights-dotnet/issues/1693) nor are we on a 2.x version of the function host (https://github.com/microsoft/ApplicationInsights-dotnet/issues/1029). We are not even explicitly installing `Microsoft.Azure.WebJobs.Logging.ApplicationInsights` in our .csproj files. The time these issues started does not line up with any deployment on our side.

- Timestamp: Started around 1:00PM Central Time on 5/17
- Function App version: ~3
- Function App name: multiple
- Region: US North Central
- Programming Language: dotnet core 3.1
- Consumption plan

Here are a few examples from Application Insights logs:

![Screen Shot 2021-05-20 at 4 57 13 PM](https://user-images.githubusercontent.com/750344/119053958-725a1b80-b98c-11eb-89d7-0863a2e42c33.png)

![Screen Shot 2021-05-20 at 4 43 46 PM](https://user-images.githubusercontent.com/750344/119052668-9157ae00-b98a-11eb-94cd-cde463b8cfcb.png)

![Screen Shot 2021-05-20 at 4 49 09 PM](https://user-images.githubusercontent.com/750344/119053175-4f7b3780-b98b-11eb-9555-b0b7e158c612.png)

**Comment by v-anvari** [comment]
Hi @MikePennington , Can you please [Share-Your-Function-App-name-privately](https://github.com/Azure/azure-functions-host/wiki/Sharing-Your-Function-App-name-privately) so that we can check the logs from our end.

**Comment by ryanbowden** [comment]
We started seeing the same thing on one of our functions around 8pm UTC UK South.

The same function is in UK West and not seeing the same issue 🤔 

@v-anvari our private function details:
2021-05-21T10:33:41.259
748227fe-7b13-48bc-9504-7c3f8782dc1f
UK South

Let me know if you need more details.

**Comment by Anusriv** [comment]
Thank you for the details @ryanbowden , We will check the logs to understand the error and revert back the findings

**Comment by v-anvari** [comment]
Thank you for the details @ryanbowden , We will check the logs to understand the error and revert back the findings

**Comment by MikePennington** [comment]
@v-anvari I don't get any "Function started" message either by triggering the function manually or by scouring the Application Insight logs. I do get a message like the one below. Does that work?

```
5/21/2021, 3:04:32.976 PM | Executing '[FUNCTION_NAME]' (Reason='This function was programmatically called via the host APIs.', Id=abd3a739-1eea-44cd-9cee-13c8d19b95dc)
```

**Comment by v-anvari** [comment]
Yes, that works, Thank you for providing the details. We will check the logs to understand the error and revert back the findings

**Comment by fasigpt** [comment]
We are hitting into the same issue starting 5/21. Our App name is "AzureSightTasks" and function that's impacted is "AzureSightDailyResourceHealthMonitorCrawler"

**Comment by edouardpoitras** [comment]
We are also having the same issue as of the 5/21.

**Comment by edouardpoitras** [comment]
A few more details that may be helpful:

- We don't use a custom TelemetryClient - just the injected ILogger object (like OP).
- Seems to happen to any function type - timer and EventGrid trigger for us.
- Happens whenever we deploy code, change function configuration, or change the App Service Plan auto-scale rules.
- Hasn't happened for our non-production App Service Plans, but they have much less traffic/activity (maybe this error is only generated for functions that are actively running during a deployment?)

Hope this helps.

**Comment by v-anvari** [comment]
Hi @MikePennington , I am unable to track your logs with the id provided, please provide any of the function names.

I see similar issues in the past as well - [https://github.com/Microsoft/ApplicationInsights-dotnet/issues/973](https://github.com/Microsoft/ApplicationInsights-dotnet/issues/973). 

We would require further details regarding your project, the repro steps, the assemblies used, code, and versions used. As there are multiple issues with the same kind off error, but different scenarios, it is important for us to understand your project specs. 

@ryanbowden , I see the following exceptions in the logs, but we are not clear,  Microsoft.Azure.WebJobs.Host.FunctionInvocationException : Exception while executing function: SessionCreatedConsumer ---> System.Exception : Could not send message Internal Webhook request to. 

Please provide the repro steps, if we can reproduce this error. 

Meanwhile, since the error is coming from App insights, please check the supported features - https://docs.microsoft.com/en-us/azure/azure-monitor/app/azure-functions-supported-features#supported-features

**Comment by johnsanmiguel** [comment]
We are also facing this issue starting from 5/26.

These are some details:

- We are using only the injected ILogger object of the function.
- This is happening to my service bus and http trigger functions.
- We are in a EP1 plan
- This is happening also in the slots

**Comment by MikePennington** [comment]
@v-anvari 

I work for a very large company and cannot share function names publicly. The information at [Sharing-Your-Function-App-name-privately](https://github.com/Azure/azure-functions-host/wiki/Sharing-Your-Function-App-name-privately) seems out of date. This is the output I get when I run the function manually in the console. Log level is set to verbose.

Filesystem logs:
```
Connected!
2021-05-28T15:49:06  Welcome, you are now connected to log-streaming service. The default timeout is 2 hours. Change the timeout with the App Setting SCM_LOGSTREAM_TIMEOUT (in seconds).
2021-05-28T15:49:06.083 [Information] Executing 'List' (Reason='This function was programmatically called via the host APIs.', Id=a6880647-77d6-431f-8d73-e9edf02b6650)
2021-05-28T15:49:07.078 [Information] Executed 'List' (Succeeded, Id=a6880647-77d6-431f-8d73-e9edf02b6650, Duration=995ms)
```

App Insights Logs:
```
Connected!
2021-05-28T15:52:13Z   [Information]   Executing 'List' (Reason='This function was programmatically called via the host APIs.', Id=3cc44228-f655-4bdf-bcf3-47b865f9f902)
```

We have many different functions that are all experiencing the same error. Here is some of the information you requested for one of them:
- Functions version: All are set to `~3`
- Assemblies used:
![image](https://user-images.githubusercontent.com/750344/120009737-e6567e00-bfa1-11eb-8cb5-ff95954e91d5.png)
- Repo steps:
  - This is difficult. As I mentioned in the OP, this error just started happening randomly on many of our functions all at the exact same time without any intervention from us. Something _clearly_ changed either in the Azure functions host or in the Applications Insights collection process to have caused this.
  - Was there any changes rolled out to either Functions host or app insights to the north-central US region around 6:00PM UTC on 5/17?

**Comment by zhenyuan0502** [comment]
I got the same problem, in my on-running production, it's started first time in 30/4/2021 region Southeast Asia, I didn't notice that when today I was trying to fix this message on local around 4 hours, nothing fixed.
![image](https://user-images.githubusercontent.com/16673159/120348937-dc2cda80-c327-11eb-876f-d5552c307ed0.png)


![image](https://user-images.githubusercontent.com/16673159/120349276-2b730b00-c328-11eb-861f-a3d4dbe9b69c.png)


I also see this message too, I have no idea to traceback.

> [LogAlways] EventSourceException while processing event "Response": System.NullReferenceException:Object reference not set to an instance of an object.
> [6/1/2021 10:40:58 PM +07:00][LogAlways] EventSourceException while processing event "Request": System.NullReferenceException:Object reference not set to an instance of an object.


![image](https://user-images.githubusercontent.com/16673159/120352138-b0f7ba80-c32a-11eb-97e4-c8d4d85bd40a.png)

**Comment by v-anvari** [comment]
Thank you for your feedback! We will investigate and update as appropriate.

**Comment by edouardpoitras** [comment]
I just want to point out that my error is exactly the same as @MikePennington and @zhenyuan0502 - Even the EventSource GUID is the same. I'm not sure how that's possible.

**Comment by zhenyuan0502** [comment]
@v-anvari 

I have reproduced the error message by using `AzureEventSourceListener` to listen `EventSource` working. What I have done is:
- Create Azure Function 3 .NET Core with Time Trigger
- Register `AzureEventSourceListener` on the function
- Call any service which is belonging to Azure-Core, I picked Search Service (simple, no need to create on Azure Portal actually)  

![image](https://user-images.githubusercontent.com/16673159/120613676-a4857600-c480-11eb-8002-629ce198e1fd.png)

My simple code:

 ```
public static class Function
    {
        [FunctionName("Function")]
        public static async Task Run([TimerTrigger("0 */5 * * * *", RunOnStartup = true)] TimerInfo myTimer, ILogger log)
        {
            log.LogInformation($"C# Timer trigger function executed at: {DateTime.Now}");

            var instance = AzureEventSourceListener.CreateTraceLogger();
            instance.EventSourceCreated += (s, e) =>
            {
                log.LogWarning($"[AzureEventSourceListencer][{e.EventSource.Name}][{e.EventSource.Guid}] Event Created", Microsoft.Extensions.Logging.LogLevel.Warning);
            };
            instance.EventWritten += (s, e) =>
            {
                log.LogWarning($"[AzureEventSourceListencer][{e.Level}] Event Written", Microsoft.Extensions.Logging.LogLevel.Warning);
            };

            AzureEventSourceListener listener = new AzureEventSourceListener(
                    (e, message) =>
                    {
                        log.LogError($"[AzureEventSourceListencer][{e.Level}] {message}", Microsoft.Extensions.Logging.LogLevel.Error);
                    }
                    , EventLevel.Error);

            await SearchService("456", "123", "my-unit-index", log);
        }

        public static async Task SearchService(string searchServiceName, string adminApiKey, string indexName, ILogger log)
        {
            Uri uri = new Uri($"https://{searchServiceName}.search.windows.net");

            var SearchIndexClient = new SearchIndexClient(uri, new AzureKeyCredential(adminApiKey));
            var SearchClient = SearchIndexClient.GetSearchClient(indexName);

            var parameters = new SearchOptions()
            {
                Size = 25
            };

            var result = await SearchClient.SearchAsync<MySearchModel>("*", parameters);

            //log.LogInformation($"Result: {result.Value.GetResults().Count()}");
        }
    }
```

Simple Search Model:

  ```
  public class MySearchModel 
    {
        [SimpleField(IsKey = true, IsFilterable = true)]
        public string Id { get; set; }

        [SearchableField(IsFilterable = true, IsFacetable = true)]
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
        public string[] Categories { get; set; }
    }
```

Nuget package:

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <AzureFunctionsVersion>v3</AzureFunctionsVersion>
    <AssemblyName>FunctionApp</AssemblyName>
    <RootNamespace>FunctionApp</RootNamespace>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Azure.Search.Documents" Version="11.2.1" />
    <PackageReference Include="Microsoft.Azure.Functions.Extensions" Version="1.1.0" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="3.0.12" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>

```
Check out reproduced project here for more information
[FunctionApp.zip](https://github.com/Azure/azure-functions-host/files/6590123/FunctionApp.zip)

Hope it helps

**Comment by vdumasacosta** [comment]
Starting to get see those when using our blob storage. Azure functions version "2.0"

**Comment by vdumasacosta** [comment]
Blob file downloading now breaks for us too same error:
Example: 
blobClient.GetBlobClient(pathName);
await blobClient.DownloadToAsync(downloadStream, cancellationToken).ConfigureAwait(false);
-> AI (Internal): [Azure-Core] ERROR: Exception in Command Processing for EventSource Azure-Core: An instance of EventSource with Guid 44cbc7c6-6776-5f3c-36c1-75cd3ef19ea9 already exists.

**Comment by jimt4593** [comment]
I started to get this issue in my Azure Function, after implementing the code found at [Log custom telemetry](https://docs.microsoft.com/en-us/azure/azure-functions/functions-dotnet-class-library?tabs=v2%2Ccmd#log-custom-telemetry-in-c-functions).   I'm only calling telemetryClient.TrackEvent(...);   

```          
	private void TrackMessageEvent(Microsoft.Graph.Message message, bool succeeded)
        {
            var processingTime = DateTime.UtcNow - message.ReceivedDateTime;
            _telemetryClient.TrackEvent(succeeded ? "Uploaded Message" : "Rejected Message",
                                        properties: new Dictionary<string, string>() { { "messageId", message.Id.GetHashCode().ToString() } },
                                        metrics: new Dictionary<string, double>() { { "processingTimeSeconds", processingTime.Value.TotalSeconds } });
        }
```

**Comment by v-anvari** [comment]
Cc @brettsam / @fabiocav

**Comment by pableess** [comment]
This same issue has happened to us twice in the last week.  Deleting and re-creating the function app seems to fix the issue for a little while.  

We are also not using any custom TelemetryClients or TelemetryConfiguration, only injected ILogger instances.

This is a major critical issue for us.  Our Queue Triggers are not executing and messages are ending up in the Poison queue because of this error.

**Comment by v-anvari** [comment]
For the team, during the error timestamps, the triggers couldn't submit the request, and eventually couldn't find the files

Few exceptions seen during the error Timestamps is as below:

FunctionInvocationId == "a6880647-77d6-431f-8d73-e9edf02b6650"
Region: North central Us
```
Timestamp : 5/18/2021 8:09:57 AM
Inner Exception Type: Azure.RequestFailedException
Total Occurrences: 21
Latest Exception Message: The specified blob does not exist. RequestId:7c62ad8e-401e-0067-31bd-4bf1b0000000 Time:2021-05-18T08:09:57.1853322Z
Status: 404 (The specified blob does not exist.)
ErrorCode: BlobNotFound

```


FunctionInvocationId == “17143ba0-9117-446a-95de-e050447ced4c”
Region: North Europe
```
Timestamp : 6/2/2021 4:13:26 PM
Inner Exception Type: FluentValidation.ValidationException
Total Occurrences: 28
Latest Exception Message: Validation failed:
-- TimesheetApproverStaffNumber: 'Timesheet Approver Staff Number' must not be empty.
```

Triaging the issue for further investigation

**Comment by mattschoutends** [comment]
We are seeing this issue as well, including the exact same EventSource GUID in the error message.  Like everyone else that has posted, we are not creating a TelemetryClient or TelemetryConfiguration.  We are using an injected ILogger.  Instances date back to 5/21 in multiple subscriptions and function application instances.

Sample recent instance:
Invocation ID:  ae22ece0-0b6c-4668-8544-96fbed1facb3
    Timestamp: 2021-06-10T15:25:57.150962Z
    Function App version: ~3
    Region: US Central
    Programming Language: dotnet core 3.1
    Consumption plan

These are timer-triggered function apps.

Package references:
<img width="743" alt="image" src="https://user-images.githubusercontent.com/53782531/121737178-26e3ea80-cabe-11eb-8148-365a0b17db0a.png">

TL;DR:  Me too!

**Comment by v-anvari** [comment]
Hi @mattschoutends , Thank you for the feedback, I can see exceptions in your function App as well, Can you provide me with the repro steps, it can help us further breakdown this error. You can send me an email

**Comment by mattschoutends** [comment]
Hi @v-anvari, emailed you some detail separately.  Like I mentioned this above, we are seeing this in multiple environments (different subscriptions, different resource groups, etc).  It happens when our function app is run manually, or from a timer trigger.  Seems to always be the same GUID in the exception message.

We have several Azure functions in our function app.  Most follow the pattern of loading data from a database, processing it, writing it to blob storage, and sending a queue message.  There is another function app that is queue-triggered.  Yet another function app operates only on blob storage.  The load/write functions and the cleanup function are on separate timer triggers, an hour apart, and we see this happen when either triggers.  

Our timer triggers are set for noon and 1PM Central.  We see these messages regularly at 12:00:01.x (where x is usually about 480ms, but varies a bit) and 1:00:00.x (same range for x).

To be completely clear about this, repro steps are as simple as letting the timer trigger the application.

**Comment by pragnagopa** [comment]
@karshinlin  - Can you take quick look  to check if this is related to https://github.com/Azure/azure-functions-host/pull/7424 ?

**Comment by anirudhgarg** [comment]
Also everyone - is this only logs noise at this point - or is it actually affecting execution.

**Comment by karshinlin** [comment]
> @karshinlin - Can you take quick look to check if this is related to #7424 ?

Don't think this relates to that change as this error was reported before that change was released. I see 2 errors in this thread:
1. ERROR: Exception in Command Processing for EventSource Azure-Core: An instance of EventSource with Guid XXX already exists.
2. ERROR: Exception in Command Processing for EventSource Azure-Core: Object reference not set to an instance of an object.
  - This seems to be caused by using an older version of Azure.Core (https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/CHANGELOG.md#key-bugs-fixed). AFAIK, the Host uses Azure.Core 1.9.0 so it should not be Host related. 

We'll need to look more into the first error though

**Comment by MikePennington** [comment]
@anirudhgarg as far as I can tell it is just noise in the logs.

**Comment by mattschoutends** [comment]
> Also everyone - is this only logs noise at this point - or is it actually affecting execution.

@anirudhgarg It seems to be just noise--as far as we can tell our functions are doing everything they are supposed to--but we stumbled across it while investigating another issue with App Insights.  

Sometimes (not every time, seems to happen most reliably the first time our function executes after a zip deployment), our function logs do not appear in App Insights.  Instead, we get "AI: Process was called on the TelemetrySink after it was disposed, the telemetry data was dropped."  One occurrence of that message appears in App Insights for each log statement we tried to emit.  (We are not disposing a TelemetrySink, and actually don't have one declared in our code.)  

They may be unrelated, in which case the paragraph above is irrelevant to this issue.  Or they may be related, in which case I hope the above paragraph is a clue.

**Comment by MikePennington** [comment]
Something that might be helpful, is that the sdkVersion is different for this error than from all the other trace logs.

![image](https://user-images.githubusercontent.com/750344/122621679-ddfed980-d05b-11eb-8baa-79b1fd9dcccd.png)

**Comment by mpaul31** [comment]
Just noticed this as well.

**Comment by attrib75** [comment]
Getting this as well

**Comment by annieEllenberger** [comment]
Getting this as well. For me, seems to be related to interacting with Blob storage SDK in combination with ILogger? Not using any custom TelemetryClient or AppInsights directly either. 

App Insights logs (same GUID as everybody above in log messages): 

![image](https://user-images.githubusercontent.com/62848777/132229645-6935d848-624e-46c3-931e-e53a5341496f.png)

Here are all of the dependencies of the entire solution: 

![image](https://user-images.githubusercontent.com/62848777/132229911-4fb25e84-aaf0-48ed-b765-ef878d5353a0.png)

**Comment by v-anvari** [comment]
Tagging @anirudhgarg / @fabiocav  for any further updates

**Comment by fabiocav** [comment]
This should be resolved when using the newest set of storage extensions. Can you  @MikePennington, can you please update the references and let us know if your problem is resolved?

**Comment by MikePennington** [comment]
@fabiocav Which nuget package specifically should be updated to resolve this? Here is an example *.csproj file from one of the projects that is having this issue.

```<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <AzureFunctionsVersion>v3</AzureFunctionsVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Azure.Identity" Version="1.2.0-preview.1" />
    <PackageReference Include="Azure.Storage.Blobs" Version="12.4.1" />
    <PackageReference Include="CsvHelper" Version="15.0.4" />
    <PackageReference Include="Microsoft.Azure.Functions.Extensions" Version="1.0.0" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="3.0.2" />
    <PackageReference Include="ReportGenerator" Version="4.4.7" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>
```

I don't have a reference to `Microsoft.Azure.WebJobs.Extensions.Storage` in there at all (other apps having this issue do have this reference though). Are you saying I need to update `Azure.Storage.Blobs` or something else?

**Comment by RohitRanjanMS** [comment]
Hi @MikePennington, the issue was with the Azure.Core nuget package. Azure.Storage.Blobs and Azure.Identity internally depends on Azure.Core. Update Azure.Storage.Blobs and Azure.Identity with the latest version.

**Comment by MikePennington** [comment]
Thanks @RohitRanjanMS I updated `Azure.Storage.Blobs` and `Azure.Identity`, deployed, and let it sit for a few days. I can confirm that these errors are gone. Thanks.

**Comment by cataggar** [comment]
I received this error in my dotnet isolated app:
```
AI (Internal): [Azure-Core] ERROR: Exception in Command Processing for EventSource Azure-Core: Object reference not set to an instance of an object.
```

after adding a dependency to:
``` xml
<PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Storage" Version="5.0.0-beta.4" />
```

I can confirm that upgrading Azure.Core worked. It was on 1.0.15, but I updated it to latest:
``` xml
<PackageReference Include="Azure.Core" Version="1.20.0" />
```

Looking forward to a 5.0.0 stable with updated dependencies.

**Comment by cataggar** [comment]
I was wrong. I'm still getting the error.
